federation/pkg/federation-controller/configmap/configmap_controller.go

```diff
@@ -0,0 +1,328 @@
+		configmapcontroller.configmapBackoff.Reset(key)
+	}
+	configmapcontroller.configmapDeliverer.DeliverAfter(key,
+		&configmapItem{namespace: namespace, name: name}, delay)
+}
+
+// Check whether all data stores are in sync. False is returned if any of the informer/stores is not yet
+// synced with the corresponding api server.
+func (configmapcontroller *ConfigMapController) isSynced() bool {
+	if !configmapcontroller.configmapFederatedInformer.ClustersSynced() {
+		glog.V(2).Infof("Cluster list not synced")
+		return false
+	}
+	clusters, err := configmapcontroller.configmapFederatedInformer.GetReadyClusters()
+	if err != nil {
+		glog.Errorf("Failed to get ready clusters: %v", err)
+		return false
+	}
+	if !configmapcontroller.configmapFederatedInformer.GetTargetStore().ClustersSynced(clusters) {
+		return false
+	}
+	return true
+}
+
+// The function triggers reconciliation of all federated configmaps.
+func (configmapcontroller *ConfigMapController) reconcileConfigMapsOnClusterChange() {
+	if !configmapcontroller.isSynced() {
+		configmapcontroller.clusterDeliverer.DeliverAt(allClustersKey, nil, time.Now().Add(configmapcontroller.clusterAvailableDelay))
+	}
+	for _, obj := range configmapcontroller.configmapInformerStore.List() {
+		configmap := obj.(*api_v1.ConfigMap)
+		configmapcontroller.deliverConfigMap(configmap.Namespace, configmap.Name, configmapcontroller.smallDelay, false)
+	}
+}
+
+func (configmapcontroller *ConfigMapController) reconcileConfigMap(namespace string, configmapName string) {
+
+	if !configmapcontroller.isSynced() {
+		configmapcontroller.deliverConfigMap(namespace, configmapName, configmapcontroller.clusterAvailableDelay, false)
+		return
+	}
+
+	key := getConfigMapKey(namespace, configmapName)
+	baseConfigMapObj, exist, err := configmapcontroller.configmapInformerStore.GetByKey(key)
+	if err != nil {
+		glog.Errorf("Failed to query main configmap store for %v: %v", key, err)
+		configmapcontroller.deliverConfigMap(namespace, configmapName, 0, true)
+		return
+	}
+
+	if !exist {
+		// Not federated configmap, ignoring.
+		return
+	}
+	baseConfigMap := baseConfigMapObj.(*api_v1.ConfigMap)
+
+	clusters, err := configmapcontroller.configmapFederatedInformer.GetReadyClusters()
+	if err != nil {
+		glog.Errorf("Failed to get cluster list: %v", err)
+		configmapcontroller.deliverConfigMap(namespace, configmapName, configmapcontroller.clusterAvailableDelay, false)
+		return
+	}
+
+	operations := make([]util.FederatedOperation, 0)
+	for _, cluster := range clusters {
+		clusterConfigMapObj, found, err := configmapcontroller.configmapFederatedInformer.GetTargetStore().GetByKey(cluster.Name, key)
+		if err != nil {
+			glog.Errorf("Failed to get %s from %s: %v", key, cluster.Name, err)
+			configmapcontroller.deliverConfigMap(namespace, configmapName, 0, true)
+			return
+		}
+
+		desiredConfigMap := &api_v1.ConfigMap{
+			ObjectMeta: util.CopyObjectMeta(baseConfigMap.ObjectMeta),
+			Data:       baseConfigMap.Data,
```

Comment:
well, not really. the copy needs to be made before any modification. If no modification is made then there is no need for deep copy. If someone is modifying the object without a deep copy then copying it here won't help - the deep copy function will get random data.